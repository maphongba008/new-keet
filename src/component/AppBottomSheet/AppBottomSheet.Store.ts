import { Keyboard } from 'react-native'
import { create } from 'zustand'
import { useShallow } from 'zustand/react/shallow'
import _debounce from 'lodash/debounce'

import {
  BACK_DEBOUNCE_DELAY,
  BACK_DEBOUNCE_OPTIONS,
  BOTTOM_SHEET_ANIMATION_DURATION,
  BOTTOM_SHEET_ANIMATION_REPLACE_TIMEOUT,
} from 'lib/constants'
import { doVibrateSuccess } from 'lib/haptics'
import { wait } from 'lib/wait'

import {
  AppSheetType,
  BottomSheetProvider,
  BottomSheetRef,
} from './AppBottomSheet.Provider'
import BottomSheetEnum from './SheetComponents/BottomSheetEnum'
import { OptionSheetOption } from './SheetComponents/components/OptionsButtonList'

interface AppBottomSheetState {
  showBottomSheet: boolean
  sheetType: number
  setSheet: (value: boolean) => void
  setSheetType: (type: number) => void
}

export const useAppBottomSheetStore = create<AppBottomSheetState>((set) => ({
  showBottomSheet: false,
  sheetType: 0,
  setSheetType: (sheetType: number) => {
    set({ sheetType })
  },
  setSheet: (value: boolean) => {
    set({ showBottomSheet: value })
  },
}))

export const useAppBottomSheetState = () =>
  useAppBottomSheetStore(useShallow((state) => state))

// Note: if you show a bottom sheet right after hide another one, make sure to wait for the animation to finish
// using setTimeout(() => {}, BOTTOM_SHEET_ANIMATION_DURATION)
export const showBottomSheet = _debounce(
  async (data: AppSheetType) => {
    doVibrateSuccess()
    Keyboard.dismiss()
    if (useAppBottomSheetStore.getState().showBottomSheet) {
      await closeBottomSheetAsync()
      await wait(BOTTOM_SHEET_ANIMATION_REPLACE_TIMEOUT)
    }

    const { bottomSheetType, bottomSheetOnDismissCallback, ...rest } = data
    // @ts-ignore
    BottomSheetProvider.bottomSheetContext = {
      bottomSheetType,
      bottomSheetOnDismissCallback,
      ...rest,
    }
    const { setSheet, setSheetType } = useAppBottomSheetStore.getState()
    setSheet(true)
    setSheetType(bottomSheetType)
  },
  BACK_DEBOUNCE_DELAY,
  BACK_DEBOUNCE_OPTIONS,
)

export const showOptionsSheet = async ({
  options,
}: {
  options: OptionSheetOption[]
}) => {
  showBottomSheet({
    bottomSheetType: BottomSheetEnum.OptionsSheet,
    options,
    bottomSheetIsTransparent: true,
  })
}

export const closeBottomSheet = () => {
  BottomSheetRef?.current?.close({
    duration: BOTTOM_SHEET_ANIMATION_DURATION,
  })
  useAppBottomSheetStore.getState().setSheet(false)
}

export const closeBottomSheetAsync = async () => {
  closeBottomSheet()
  await wait(
    BOTTOM_SHEET_ANIMATION_DURATION + BOTTOM_SHEET_ANIMATION_REPLACE_TIMEOUT,
  )
}
