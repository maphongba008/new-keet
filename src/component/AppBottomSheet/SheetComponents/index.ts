import { lazy } from 'react'

import { withSuspense } from 'lib/hoc/withSuspense'

import BottomSheetEnum from './BottomSheetEnum'
import MessageOptionsSheet from './MessageOptionSheet'
import RoomActionSheet from './RoomActionSheet'

const CallAutoEndPopup = withSuspense(lazy(() => import('./CallAutoEndPopup')))
const CallStarted = withSuspense(lazy(() => import('./CallStarted')))
const ChatEventDeleteConfirmSheet = withSuspense(
  lazy(() => import('./ChatEventDeleteConfirmSheet')),
)
const ChatEventOptionsSheet = withSuspense(
  lazy(() => import('./ChatEventOptionsSheet')),
)
const ChatLinkConfirmation = withSuspense(
  lazy(() => import('./ChatLinkConfirmation')),
)
const CleanDevice = withSuspense(lazy(() => import('./CleanDevice')))
const ClearAllFiles = withSuspense(lazy(() => import('./ClearAllFiles')))
const CommunityContacts = withSuspense(
  lazy(() => import('./CommunityContacts')),
)
const ConfirmBlockDialog = withSuspense(
  lazy(() => import('./ConfirmBlockDialog')),
)
const ConfirmDialog = withSuspense(lazy(() => import('./ConfirmDialog')))
const ConfirmInappropriateMessageDialog = withSuspense(
  lazy(() => import('./ConfirmInappropriateMessageDialog')),
)
const DeviceDetailsSheet = withSuspense(
  lazy(() => import('./DeviceDetailsSheet')),
)
const Dialog = withSuspense(lazy(() => import('./Dialog')))
const DMRequest = withSuspense(lazy(() => import('./DMRequest')))
const EmojiSourceSheet = withSuspense(lazy(() => import('./EmojiSourceSheet')))
const FeedbackForm = withSuspense(lazy(() => import('./FeedbackForm')))
const IdentityBackupConfirmSheet = withSuspense(
  lazy(() => import('./IdentityBackupConfirmSheet')),
)
const InviteTypeOptionsSheet = withSuspense(
  lazy(() => import('./InviteTypeOptionsSheet')),
)
const ListOptionsSheet = withSuspense(lazy(() => import('./ListOptionsSheet')))
const NetworkBottomSheet = withSuspense(
  lazy(() => import('./NetworkBottomSheet')),
)
const OptionsSheet = withSuspense(lazy(() => import('./OptionsSheet')))
const ParticipantsSheet = withSuspense(
  lazy(() => import('./ParticipantsSheet')),
)
const PermissionRequired = withSuspense(
  lazy(() => import('./PermissionRequired')),
)
const RecoverWalletSheet = withSuspense(
  lazy(() => import('./RecoverWalletSheet')),
)
const RoleChangeConfirm = withSuspense(
  lazy(() => import('./RoleChangeConfirm')),
)
const RoomAdminReset = withSuspense(lazy(() => import('./RoomAdminReset')))
const RoomAvatarBottomSheet = withSuspense(lazy(() => import('./RoomAvatar')))
const RoomInviteType = withSuspense(lazy(() => import('./RoomInvitationInfo')))
const RoomOptionsStats = withSuspense(lazy(() => import('./RoomOptionsStats')))
const TurnOffPasscode = withSuspense(lazy(() => import('./TurnOffPasscode')))
const WalletDelete = withSuspense(lazy(() => import('./WalletDelete')))
const WalletSettings = withSuspense(lazy(() => import('./WalletSettings')))
const ChannelRoomTips = withSuspense(
  lazy(() => import('screen/CreateChannel/ChannelRoomTips')),
)

export default {
  [BottomSheetEnum.MessageOptionsSheet]: MessageOptionsSheet,
  [BottomSheetEnum.ConfirmDialog]: ConfirmDialog,
  [BottomSheetEnum.ConfirmBlockDialog]: ConfirmBlockDialog,
  [BottomSheetEnum.ConfirmInappropriateMessageDialog]:
    ConfirmInappropriateMessageDialog,
  [BottomSheetEnum.RecoverWalletSheet]: RecoverWalletSheet,
  [BottomSheetEnum.ParticipantsSheet]: ParticipantsSheet,
  [BottomSheetEnum.OptionsSheet]: OptionsSheet,
  [BottomSheetEnum.CallStarted]: CallStarted,
  [BottomSheetEnum.RoomActionSheet]: RoomActionSheet,
  [BottomSheetEnum.NetworkBottomSheet]: NetworkBottomSheet,
  [BottomSheetEnum.ChatEventOptionsSheet]: ChatEventOptionsSheet,
  [BottomSheetEnum.PermissionRequired]: PermissionRequired,
  [BottomSheetEnum.RoomInviteType]: RoomInviteType,
  [BottomSheetEnum.ClearAllFiles]: ClearAllFiles,
  [BottomSheetEnum.Dialog]: Dialog,
  [BottomSheetEnum.FeedbackFormSheet]: FeedbackForm,
  [BottomSheetEnum.EmojiSourceSheet]: EmojiSourceSheet,
  [BottomSheetEnum.InviteTypeOptionsSheet]: InviteTypeOptionsSheet,
  [BottomSheetEnum.RoomAvatarBottomSheet]: RoomAvatarBottomSheet,
  [BottomSheetEnum.ChannelRoomTips]: ChannelRoomTips,
  [BottomSheetEnum.IdentityBackupConfirmSheet]: IdentityBackupConfirmSheet,
  [BottomSheetEnum.CleanDeviceBottomSheet]: CleanDevice,
  [BottomSheetEnum.TurnOffPasscodeSheet]: TurnOffPasscode,
  [BottomSheetEnum.ChatLinkConfirmationSheet]: ChatLinkConfirmation,
  [BottomSheetEnum.RoomOptionsStats]: RoomOptionsStats,
  [BottomSheetEnum.CommunityContacts]: CommunityContacts,
  [BottomSheetEnum.RoomAdminReset]: RoomAdminReset,
  [BottomSheetEnum.DeviceDetailsBottomSheet]: DeviceDetailsSheet,
  [BottomSheetEnum.DMRequest]: DMRequest,
  [BottomSheetEnum.ChatEventDeleteConfirmSheet]: ChatEventDeleteConfirmSheet,
  [BottomSheetEnum.WalletSettings]: WalletSettings,
  [BottomSheetEnum.WalletDelete]: WalletDelete,
  [BottomSheetEnum.RoleChangeConfirm]: RoleChangeConfirm,
  [BottomSheetEnum.ListOptionsSheet]: ListOptionsSheet,
  [BottomSheetEnum.CallAutoEndPopup]: CallAutoEndPopup,
}
