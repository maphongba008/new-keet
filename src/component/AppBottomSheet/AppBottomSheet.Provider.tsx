import React from 'react'
import { type BottomSheetModal } from '@gorhom/bottom-sheet'
import { BackdropPressBehavior } from '@gorhom/bottom-sheet/lib/typescript/components/bottomSheetBackdrop/types'

import { ChannelRoomTipsProps } from 'screen/CreateChannel/ChannelRoomTips'

// Enum for bottom sheet types
import BottomSheetEnum from './SheetComponents/BottomSheetEnum'
// Import props/interfaces from various components
import { CallStartedInterface } from './SheetComponents/CallStarted'
import { ChatEventDeleteConfirmSheetI } from './SheetComponents/ChatEventDeleteConfirmSheet'
import { ChatEventOptionsSheetProp } from './SheetComponents/ChatEventOptionsSheet'
import { ChatLinkConfirmationInterface } from './SheetComponents/ChatLinkConfirmation'
import { ClearAllFilesInterface } from './SheetComponents/ClearAllFiles'
import { ConfirmBlockDialogInterface } from './SheetComponents/ConfirmBlockDialog'
import { ConfirmDialogInterface } from './SheetComponents/ConfirmDialog'
import { ConfirmInappropriateMessageDialogInterface } from './SheetComponents/ConfirmInappropriateMessageDialog'
import { DeviceDetailsInterface } from './SheetComponents/DeviceDetailsSheet'
import { DialogInterface } from './SheetComponents/Dialog'
import { DMRequestInterface } from './SheetComponents/DMRequest'
import { EmojiSourceSheetInterface } from './SheetComponents/EmojiSourceSheet'
import { FeedbackFormSheetInterface } from './SheetComponents/FeedbackForm'
import { IdentityBackupConfirmSheetInterface } from './SheetComponents/IdentityBackupConfirmSheet'
import { InviteTypeOptionsSheetProps } from './SheetComponents/InviteTypeOptionsSheet'
import { ListOptionSheetPropsI } from './SheetComponents/ListOptionsSheet'
import { MessageOptionsSheetInterface } from './SheetComponents/MessageOptionSheet'
import { OptionSheetProps } from './SheetComponents/OptionsSheet'
import { ParticipantsSheetProps } from './SheetComponents/ParticipantsSheet'
import { PermissionRequiredInterface } from './SheetComponents/PermissionRequired'
import { RoleChangeConfirmInterface } from './SheetComponents/RoleChangeConfirm'
import { RoomAdminResetInterface } from './SheetComponents/RoomAdminReset'
import { RoomAvatarSheetProps } from './SheetComponents/RoomAvatar'
import { RoomOptionsStatsI } from './SheetComponents/RoomOptionsStats'

// props specific for bottom sheet
interface BaseBottomSheetProps {
  bottomSheetIsTransparent?: boolean
  bottomSheetOnDismissCallback?: () => void
  bottomSheetBackdropPressBehaviour?: BackdropPressBehavior
  bottomSheetEnablePanDownToClose?: boolean
}

// props specific for bottom sheet's rendered component
type BottomSheetComponentProps = {
  [BottomSheetEnum.MessageOptionsSheet]: MessageOptionsSheetInterface
  [BottomSheetEnum.ConfirmDialog]: ConfirmDialogInterface
  [BottomSheetEnum.ConfirmBlockDialog]: ConfirmBlockDialogInterface
  [BottomSheetEnum.ConfirmInappropriateMessageDialog]: ConfirmInappropriateMessageDialogInterface
  [BottomSheetEnum.InviteTypeOptionsSheet]: InviteTypeOptionsSheetProps
  [BottomSheetEnum.RecoverWalletSheet]: {}
  [BottomSheetEnum.ParticipantsSheet]: ParticipantsSheetProps
  [BottomSheetEnum.OptionsSheet]: OptionSheetProps
  [BottomSheetEnum.CallStarted]: CallStartedInterface
  [BottomSheetEnum.RoomActionSheet]: {}
  [BottomSheetEnum.NetworkBottomSheet]: {}
  [BottomSheetEnum.ChatEventOptionsSheet]: ChatEventOptionsSheetProp
  [BottomSheetEnum.PermissionRequired]: PermissionRequiredInterface
  [BottomSheetEnum.RoomInviteType]: {}
  [BottomSheetEnum.ClearAllFiles]: ClearAllFilesInterface
  [BottomSheetEnum.ChatLinkConfirmationSheet]: ChatLinkConfirmationInterface
  [BottomSheetEnum.Dialog]: DialogInterface
  [BottomSheetEnum.FeedbackFormSheet]: FeedbackFormSheetInterface
  [BottomSheetEnum.EmojiSourceSheet]: EmojiSourceSheetInterface
  [BottomSheetEnum.RoomAvatarBottomSheet]: RoomAvatarSheetProps
  [BottomSheetEnum.ChannelRoomTips]: ChannelRoomTipsProps
  [BottomSheetEnum.IdentityBackupConfirmSheet]: IdentityBackupConfirmSheetInterface
  [BottomSheetEnum.CleanDeviceBottomSheet]: {}
  [BottomSheetEnum.TurnOffPasscodeSheet]: {}
  [BottomSheetEnum.RoomOptionsStats]: RoomOptionsStatsI
  [BottomSheetEnum.CommunityContacts]: {}
  [BottomSheetEnum.RoomAdminReset]: RoomAdminResetInterface
  [BottomSheetEnum.DeviceDetailsBottomSheet]: DeviceDetailsInterface
  [BottomSheetEnum.DMRequest]: DMRequestInterface
  [BottomSheetEnum.ChatEventDeleteConfirmSheet]: ChatEventDeleteConfirmSheetI
  [BottomSheetEnum.WalletSettings]: {}
  [BottomSheetEnum.WalletDelete]: {}
  [BottomSheetEnum.RoleChangeConfirm]: RoleChangeConfirmInterface
  [BottomSheetEnum.ListOptionsSheet]: ListOptionSheetPropsI
  [BottomSheetEnum.CallAutoEndPopup]: {}
}

type BottomSheetProps = {
  [K in keyof BottomSheetComponentProps]: {
    bottomSheetType: K
  } & BottomSheetComponentProps[K]
}[keyof BottomSheetComponentProps]

// All types are combined here so that when showing a sheet, you only need to provide the enum.
// Call showBottomSheet({ bottomSheetType: BottomSheetEnum }) and the appropriate interface/type will be inferred automatically.
export type AppSheetType = BottomSheetProps & BaseBottomSheetProps

class BottomSheetProviderClass {
  bottomSheetContext: AppSheetType
  constructor() {
    this.bottomSheetContext = {
      bottomSheetType: BottomSheetEnum.RoomActionSheet, // Default type
    }
  }
}

export const BottomSheetProvider = new BottomSheetProviderClass()

export const BottomSheetRef = React.createRef<BottomSheetModal>()
