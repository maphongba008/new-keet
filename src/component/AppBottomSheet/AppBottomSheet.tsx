import React, { memo, useCallback, useEffect, useMemo } from 'react'

import { BottomSheetProvider, BottomSheetRef } from './AppBottomSheet.Provider'
import {
  closeBottomSheet,
  showBottomSheet,
  useAppBottomSheetStore,
} from './AppBottomSheet.Store'
import SheetComponents from './SheetComponents/index'
import { BottomSheetBase } from '../BottomSheetBase'

const AppBottomSheet = () => {
  const {
    bottomSheetType,
    bottomSheetIsTransparent,
    bottomSheetBackdropPressBehaviour,
    bottomSheetEnablePanDownToClose,
    bottomSheetOnDismissCallback,
    ...rest
  } = BottomSheetProvider.bottomSheetContext

  useEffect(() => {
    const timeout = setTimeout(() => BottomSheetRef?.current?.present())
    return () => clearTimeout(timeout)
  }, [])

  const onBottomSheetClose = useCallback(() => {
    if (bottomSheetOnDismissCallback) {
      bottomSheetOnDismissCallback()
    }
    closeBottomSheet()
  }, [bottomSheetOnDismissCallback])

  const renderComponent = useMemo(() => {
    const Component = (SheetComponents[bottomSheetType] as any) || null
    return <Component {...rest} />
  }, [bottomSheetType, rest])

  return (
    <BottomSheetBase
      sheetRef={BottomSheetRef}
      isTransparent={bottomSheetIsTransparent}
      backdropPressBehaviour={bottomSheetBackdropPressBehaviour ?? 'close'}
      enablePanDownToClose={bottomSheetEnablePanDownToClose ?? true}
      onClose={onBottomSheetClose}
    >
      {renderComponent}
    </BottomSheetBase>
  )
}

export default memo(AppBottomSheet)
export { showBottomSheet, closeBottomSheet, useAppBottomSheetStore }
