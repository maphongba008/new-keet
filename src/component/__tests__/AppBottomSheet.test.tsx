import React from 'react'
import { render, screen } from '@testing-library/react-native'
import { BackHandler } from 'react-native'

import AppBottomSheet from 'component/AppBottomSheet/AppBottomSheet'
import { BottomSheetProvider } from 'component/AppBottomSheet/AppBottomSheet.Provider'

const BOTTOM_SHEET_TYPE = 10

jest.mock('../AppBottomSheet/AppBottomSheet.Provider', () => ({
  BottomSheetProvider: {
    bottomSheetContext: {
      bottomSheetIsTransparent: false,
      bottomSheetType: BOTTOM_SHEET_TYPE,
    },
  },
}))

jest.mock('react-native-safe-area-context', () => ({
  ...jest.requireActual('react-native-safe-area-context'),
  useSafeAreaInsets: jest.fn().mockReturnValue({
    bottom: 0,
  }),
}))

jest.mock('@gorhom/bottom-sheet', () => {
  const reactNative = jest.requireActual('react-native')
  const { View } = reactNative

  return {
    __esModule: true,
    default: View,
    BottomSheetModal: View,
    BottomSheetView: View,
    BottomSheetModalProvider: View,
    BottomSheetTextInput: View,
    useBottomSheetModal: () => ({
      present: () => {},
      dismiss: () => {},
    }),
  }
})

jest.mock('react-native-reanimated')
jest.mock('../AppBottomSheet/SheetComponents', () => {
  const SheetComponent = require('../__mocks__/BottomSheetComponent').default
  return {
    __esModule: true,
    default: {
      [BOTTOM_SHEET_TYPE]: SheetComponent,
    },
  }
})

describe('App bottom sheet works correctly', () => {
  it('should render sheet content when we provide valid bottomSheetType', async () => {
    render(<AppBottomSheet />)
    await screen.findAllByText('Test Component')
    expect(screen.toJSON()).toMatchSnapshot()
  })
  it('should have style set when isTransparent is false', () => {
    render(<AppBottomSheet />)
    expect(screen.getByTestId('bottom_sheet_view')).toHaveStyle({
      backgroundColor: '#151823',
      flexDirection: 'column',
      justifyContent: 'flex-start',
      padding: 16,
    })
  })
  it('should have style set when isTransparent is true', () => {
    const bottomSheetContext = BottomSheetProvider.bottomSheetContext
    bottomSheetContext.bottomSheetIsTransparent = true
    render(<AppBottomSheet />)
    expect(screen.getByTestId('bottom_sheet_view')).toHaveStyle({
      paddingBottom: 16,
    })
    expect(screen.toJSON()).toMatchSnapshot()
  })
  // TODO: fix the workaround
  xit('should close sheet when user swipes back', () => {
    const indexRef = 2
    jest.spyOn(React, 'useRef').mockReturnValue({ current: indexRef })
    const addEventListenerMocked = jest.spyOn(
      BackHandler,
      'addEventListener',
    ) as jest.SpyInstance<any>
    render(<AppBottomSheet />)
    const backAction = addEventListenerMocked.mock.calls[0][1]
    const result = backAction()
    expect(result).toBeTruthy()
  })
})
