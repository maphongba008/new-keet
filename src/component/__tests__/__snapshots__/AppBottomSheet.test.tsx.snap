// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App bottom sheet works correctly should have style set when isTransparent is true 1`] = `
<View
  backdropComponent={[Function]}
  backgroundStyle={
    {
      "backgroundColor": "transparent",
    }
  }
  enableContentPanningGesture={false}
  enableDismissOnClose={true}
  enableDynamicSizing={true}
  enablePanDownToClose={true}
  handleIndicatorStyle={
    {
      "height": 0,
    }
  }
  index={0}
  keyboardBlurBehavior="restore"
  onChange={[Function]}
  onDismiss={[Function]}
  style={
    {
      "marginTop": 0,
    }
  }
>
  <View
    style={
      {
        "paddingBottom": 16,
        "paddingHorizontal": 12,
      }
    }
    testID="bottom_sheet_view"
  >
    <View
      testID="test_bottom_sheet"
    >
      <Text>
        Test Component
      </Text>
    </View>
  </View>
</View>
`;

exports[`App bottom sheet works correctly should render sheet content when we provide valid bottomSheetType 1`] = `
<View
  backdropComponent={[Function]}
  backgroundStyle={
    {
      "backgroundColor": "#151823",
    }
  }
  enableContentPanningGesture={false}
  enableDismissOnClose={true}
  enableDynamicSizing={true}
  enablePanDownToClose={true}
  handleIndicatorStyle={
    {
      "backgroundColor": "#ffffff",
    }
  }
  index={0}
  keyboardBlurBehavior="restore"
  onChange={[Function]}
  onDismiss={[Function]}
  style={
    {
      "marginTop": 0,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#151823",
          "flexDirection": "column",
          "justifyContent": "flex-start",
          "padding": 16,
        },
        {
          "paddingBottom": 16,
        },
        undefined,
      ]
    }
    testID="bottom_sheet_view"
  >
    <View
      testID="test_bottom_sheet"
    >
      <Text>
        Test Component
      </Text>
    </View>
  </View>
</View>
`;
